lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+				branch.getOperation()->print(llvm::outs());
+				llvm::outs()<<"\nregion from: ";
+				if(auto regionf=regionFrom.getRegionOrNull()){
+					regionf->front().print(llvm::outs());
+				}
+				else
+					llvm::outs()<<"null\n";
+				llvm::outs()<<"region to: ";
+				if(auto regiont=regionTo.getRegionOrNull()){
+					regiont->front().print(llvm::outs());
+				}
+				else
+					llvm::outs()<<"null\n";
+				auto res1=after->setCurrentBranch(&branch);
+				//auto res1=mlir::ChangeResult::Change;
+				auto res2=after->join(before);
+				auto res= res1==mlir::ChangeResult::Change ? res1 : (res2==mlir::ChangeResult::Change ? res2 : mlir::ChangeResult::NoChange);
+				
+				// Let's try this to have a bit of fun -> ChangeResult is an enum (this should not work
+				propagateIfChanged(after,res);
+			}
+
+			// At the end we can take the information about the arguments in our lattice and attach it
+			// to the function arguments
+			// TODO: understand when this is called and where to append the information obtained when finishing
+			// For the moment this function is just observing the IR, so we can leave it as it is
+			void setToExitState(ConstraintsLattice* lattice) override
+			{
+				
+				llvm::outs()<<"something was set to exit state\n";
+				
+			}
+
+			public:
+
+			void printRanges(mlir::Operation* op){
+				auto* lattice = op != nullptr ? getLattice(mlir::ProgramPoint(op))
+																			: getLattice(op->getBlock());
+				llvm::outs()<<"KEK ";
+				lattice->print(llvm::outs());
+			}
+
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getUnderlyingLattice(mlir::Operation* op){
+				return this->getLattice(mlir::ProgramPoint(op))->getUnderlyingLattice();
+			}
+		};
+
+		// TODO: should I put it in another namespace ? (as in "ActionLiveness.cpp")
+		// This is the class that starts the analysis (I guess)
+		class Constraints
+		{
+			public:
+			explicit Constraints(mlir::rlc::FlatFunctionOp fun)
+					: function(fun), config()
+			{
+				config.setInterprocedural(false);
+				solver = std::make_unique<DataFlowSolver>(config);
+				//TODO: understand what are these -> other analysis? so should I keep them ? I think I should
+				solver->load<mlir::dataflow::DeadCodeAnalysis>();
+				solver->load<mlir::dataflow::SparseConstantPropagation>();
+				analysis = solver->load<ConstraintsAnalysis>(symbolTable);
+				auto top = fun;
+				llvm::outs()<<"Initialize analysis from operation: ";
+				top->print(llvm::outs());
+				auto res = solver->initializeAndRun(top);
+			}
+			
+			void printRangesFound(mlir::Operation* op){
+				llvm::outs()<<"Printing operation: ";
+				op->print(llvm::outs());
+				llvm::outs()<<"\n";
+				this->analysis->printRanges(op);
+			}
+
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getOperationLattice(mlir::Operation* op){
+				return this->analysis->getUnderlyingLattice(op);
+			}
+
+			private:
+			mlir::SymbolTableCollection symbolTable;
```

Comment:
why do we need this? i don't see it being accessed anywhere. Was it just something the particular type of analysis require?

Reply:
Actually it is used some lines above, but I am not sure if it is needed or not since it is there from the first analysis that I copied.